pkg/storage/etcd3/watch_test.go

```diff
@@ -0,0 +1,175 @@
+			err := store.GuaranteedUpdate(ctx, tt.key, out, true, nil, storage.SimpleUpdate(
+				func(runtime.Object) (runtime.Object, error) {
+					return watchTest.obj, nil
+				}))
+			if err != nil {
+				t.Fatalf("GuaranteedUpdate failed: %v", err)
+			}
+			if watchTest.expectEvent {
+				testCheckResult(t, i, watch.Added, w, out)
+			}
+		}
+		w.Stop()
+		testCheckStop(t, i, w)
+	}
+}
+
+// TestWatchSyncAdd tests that if some object has been set on a key,
+// watch from 0 should sync up, grab the object, and return it in ResultChan().
+func TestWatchSyncAdd(t *testing.T) {
+	ctx, store, cluster := testSetup(t)
+	defer cluster.Terminate(t)
+	key, storedObj := testPropogateStore(t, store, ctx, &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}})
+
+	w, err := store.Watch(ctx, key, "0", storage.Everything)
+	if err != nil {
+		t.Fatalf("Watch failed: %v", err)
+	}
+	testCheckResult(t, 0, watch.Added, w, storedObj)
+	w.Stop()
+	testCheckStop(t, 0, w)
+}
+
+func TestWatchListAdd(t *testing.T) {
+	ctx, store, cluster := testSetup(t)
+	defer cluster.Terminate(t)
+
+	podFoo := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
+	podBar := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "bar"}}
+
+	tests := []struct {
+		key        string
+		filter     storage.FilterFunc
+		watchTests []*testWatchStruct
+	}{{ // create a key
+		key:        "/somekey-1",
+		watchTests: []*testWatchStruct{{podFoo, true}},
+		filter:     storage.Everything,
+	}, { // create a key but obj gets filtered
+		key:        "/somekey-2",
+		watchTests: []*testWatchStruct{{podFoo, false}},
+		filter:     func(runtime.Object) bool { return false },
+	}, { // create a key but obj gets filtered. Then update it and not filtered the second time
+		key:        "/somekey-3",
+		watchTests: []*testWatchStruct{{podFoo, false}, {podBar, true}},
+		filter: func(obj runtime.Object) bool {
+			pod := obj.(*api.Pod)
+			return pod.Name == "bar"
+		},
+	}}
+	for i, tt := range tests {
+		w, err := store.WatchList(ctx, tt.key, "0", tt.filter)
+		if err != nil {
+			t.Fatalf("Watch failed: %v", err)
+		}
+		for _, watchTest := range tt.watchTests {
+			out := &api.Pod{}
+			err := store.GuaranteedUpdate(ctx, tt.key+"/item", out, true, nil, storage.SimpleUpdate(
+				func(runtime.Object) (runtime.Object, error) {
+					return watchTest.obj, nil
+				}))
+			if err != nil {
+				t.Fatalf("GuaranteedUpdate failed: %v", err)
+			}
+			if watchTest.expectEvent {
+				testCheckResult(t, i, watch.Added, w, out)
+			}
+		}
+		w.Stop()
+		testCheckStop(t, i, w)
+	}
+}
+
+func testCheckResult(t *testing.T, i int, expectEventType watch.EventType, w watch.Interface, expectObj *api.Pod) {
+	select {
+	case res := <-w.ResultChan():
+		if res.Type != expectEventType {
+			t.Errorf("#%d: event type want=%v, get=%v", i, expectEventType, res.Type)
+		}
+		if !reflect.DeepEqual(expectObj, res.Object) {
+			t.Errorf("#%d: obj want=\n%#v\nget=\n%#v", i, expectObj, res.Object)
+		}
+	case <-time.After(1 * time.Second):
```

Comment:
We used to have some problems with timeouts in the past (due to overloaded machines etc.), so we are trying now to be less conservative.
Please use "wait.ForeverTestTimeout" instead